sycl/test/e2e_test_requirements/check-correctness-of-requires.py

```diff
@@ -0,0 +1,116 @@
+# See check-correctness-of-requires.cpp
+
+import re
+import sys
+
+
+# To parse .def files to get aspects and architectures
+def parse_defines(path, macro, prefix):
+    features = set()
+    with open(path, "r") as file:
+        for line in file:
+            if line.startswith(macro):
+                feature = line.split("(")[1].split(",")[0].strip()
+                features.add(f"{prefix}-{feature}")
+        return features
+
+
+def parse_requirements(input_data_path, sycl_include_dir_path):
+    available_features = {
```

Comment:
We must import features from llvm/sycl/test-e2e/lit.cfg.py instead of building a manually maintained copy of this list.

Reply:
> We must import features from llvm/sycl/test-e2e/lit.cfg.py instead of building a manually maintained copy of this list.

Can we really do that in a reliable manner? Those features depend on the environment (OS, available tools, build configuration, enabled projects, etc.).

Reply:
> We must import features from llvm/sycl/test-e2e/lit.cfg.py instead of building a manually maintained copy of this list.

+1 to Alexey's question. lit.cfg.py generates the set of available features "online" according to the specific env. Do you know if it's possible to get the whole set regardless of env? I agree it's not the best idea to maintain this set manually, but I'm not sure if we have another options.

Reply:
I have an idea, but it requires refactoring of llvm/sycl/test-e2e/lit.cfg.py. 
We can have a dictionary of all features with callbacks like this:
```python
features = {
'x': available, # 'available' function just returns true, i.e. make 'x' always available.
'y': is_y_available, # 'is_y_available' function returns true if feature 'y' is available.
'z': available, # 'available' function just returns true
...
}
```

1. To get the list of all features that can be used by e2e tests we just need to get the keys (i.e. `features.keys()`)
2. To build the list of available features, we just go over all the items in the dictionary and use functions to decide which features are available.

The benefit of this implementation, you don't need to update your test when new feature is added.

Reply:
Are there always available features?
I like that the suggested approach is more "centralized", but I guess it will take some time as it requires some further discussion. I'd prefer to merge this PR as is, since I don't think we get new features really often. Then we can proceed with a follow-up patch to move this set to lit.cfg.py.

Reply:
> we get new features really often.

Considering that, do we need to rush with merging this solution? I suggest we improve the solution before merging.

Reply:
I'm trying to prototype this and have a few questions:

1. Do we really need to proceed with such dictionary and not the set? Such dictionary requires the developer to not only add a key for new feature, but also a function to detect the availability of this feature.
2. It leads to dozens of these "detect" functions.
3. And it looks like then we can't auto generate aspect and architecture features (I mean as it's done now in this patch), as they may require some specific approach to detect them.

In total from my POV the idea of such dictionary looks too complicated. Looks like it requires a thorough refactoring now and a lot of work in future to support it, however, I don't see much benefit from this.

I suggest to keep it as set and move it to `lif.cfg.py` as function, so we can call it from this test. Something like:
```py
def get_all_features:
    all_features {
        "cpu",
        "gpu",
        ...
    }
    all_features.update(generate_aspects())
    all_features.update(generate_archs())
    return all_features
```
What do you think?

Reply:
> I'm trying to prototype this and have a few questions:
> 
> 1. Do we really need to proceed with such dictionary and not the set? Such dictionary requires the developer to not only add a key for new feature, but also a function to detect the availability of this feature.

The main benefit of using dictionary is that it is a single source of all features. What I'd like to avoid is to build two separate lists of features: all possible features and all available features. The idea with dictionary allows us to build "all available features" list automatically. Current proposal requires developers to add features into two lists separately and it's easy to miss. I'm open to alternative ideas, which will guarantee that if developer adds new llvm-lit feature to the list of "available features", there is no way to skip the feature in the list of "all possible features".

Your test is kind of guarantees that to some extent i.e. it's supposed to fail if e2e tests will use some llvm-lit feature `X`, which is not in the list of `available_features` defined by sycl/test/e2e_test_requirements/check-correctness-of-requirements.py. Am I right?
 
> 2. It leads to dozens of these "detect" functions.

Right. Is that a problem? We have all this logic already. It's just not structured as separate "detect" functions.

> 3. And it looks like then we can't auto generate aspect and architecture features (I mean as it's done now in this patch), as they may require some specific approach to detect them.

Why not?

> 
> In total from my POV the idea of such dictionary looks too complicated. Looks like it requires a thorough refactoring now and a lot of work in future to support it, however, I don't see much benefit from this.

I agree that it requires refactoring now, but I hope you see the benefit for future changes, which I tried to explain above. If you have any better ideas, I'm all ears. 

> 
> I suggest to keep it as set and move it to `lif.cfg.py` as function, so we can call it from this test. Something like:
> 
> ```python
> def get_all_features:
>     all_features {
>         "cpu",
>         "gpu",
>         ...
>     }
>     all_features.update(generate_aspects())
>     all_features.update(generate_archs())
>     return all_features
> ```
> 
> What do you think?

How is this different from your current patch?

Reply:
> Current proposal requires developers to add features into two lists separately and it's easy to miss.

If a feature added as available, but left unused - it doesn't matter if we list it in possible features and it is likely a user error, which doesn't sound too serious, we can ignore that. For example, we have `FileCheck` registered as a feature, but I doubt that we use it anywhere.

> which will guarantee that if developer adds new llvm-lit feature to the list of "available features", there is no way to skip the feature in the list of "all possible features".

If a feature added as available, adds it use, but isn't registered as "known", this new test will fail, notifying a PR author.

Therefore, I don't think that we can miss anything important when we adding features.

However, if a feature was removed from available features, but kept in "known" - we won't highlight any incorrect uses of it. From that point of view, the approach proposed by @bader is better, I think.
And we do remove features from time to time - the most often used example is renaming of aspects added by experimental extensions.

> It leads to dozens of these "detect" functions.
> Right. Is that a problem? We have all this logic already. It's just not structured as separate "detect" functions.

I think it worth exploring that path. `lit.cfg.py` is already quite lengthy and I wonder if we can structure it better. I don't think that we really need "detect" functions, but we probably need a helper functions to register features. It would accept two arguments: a feature itself and a boolean saying whether its available. Under the hood it will unconditionally register a feature as known, but conditionally register it as available.

```python
if platform.system() == "Linux":
    config.available_features.add("linux")
    llvm_config.with_system_environment(["LD_LIBRARY_PATH", "LIBRARY_PATH", "CPATH"])
    llvm_config.with_environment(
        "LD_LIBRARY_PATH", config.sycl_libs_dir, append_path=True
    )
``` 

Would turn into:

```python
if platform.system() == "Linux":
    register_feature(config, "linux", True)
    llvm_config.with_system_environment(["LD_LIBRARY_PATH", "LIBRARY_PATH", "CPATH"])
    llvm_config.with_environment(
        "LD_LIBRARY_PATH", config.sycl_libs_dir, append_path=True
    )

# For simpler cases:
register_feature(config, "my-new-os", platform.system() == "New OS")
```

And that doesn't look that bad. We have plenty of features which are detected by running an external process - there is plenty of duplicated code and I would love to see some refactoring there to have less repetitions.

Reply:
I'm okay with registering function. 

One very minor concern I have is using config across `lit.cfg.py`. I suppose we can delay updating the list of available features until all features are registered. Another approach is to create a wrapper object containing config and make register function a method of this function object. If it's not too much trouble it would be nice to avoid using config directly.